server/src/main/java/org/opensearch/cluster/routing/OperationRouting.java

```diff
@@ -343,13 +349,70 @@ private ShardIterator preferenceActiveShardIterator(
         routingHash = 31 * routingHash + indexShard.shardId.hashCode();
-        if (ignoreAwarenessAttributes()) {
+        if (weightedRoutingMetadata != null && weightedRoutingMetadata.getWeightedRouting().isSet() && isStrictWeightedShardRouting) {
+            return indexShard.activeInitializingShardsSimpleWeightedIt(
+                weightedRoutingMetadata.getWeightedRouting(),
+                nodes,
+                getWeightedRoutingDefaultWeight(),
+                routingHash
+            );
+        } else if (ignoreAwarenessAttributes()) {
             return indexShard.activeInitializingShardsIt(routingHash);
         } else {
             return indexShard.preferAttributesActiveInitializingShardsIt(awarenessAttributes, nodes, routingHash);
         }
     }
 
+    private ShardIterator preferenceSearchWithWeightedRouting(
+        IndexShardRoutingTable indexShard,
+        String localNodeId,
+        DiscoveryNodes nodes,
+        String preference,
+        Preference preferenceType,
+        WeightedRoutingMetadata weightedRoutingMetadata
+    ) {
+        switch (preferenceType) {
+            case PREFER_NODES:
+                final Set<String> nodesIds = Arrays.stream(preference.substring(Preference.PREFER_NODES.type().length() + 1).split(","))
+                    .collect(Collectors.toSet());
+                return indexShard.preferNodeActiveInitializingShardsWeightedIt(
+                    nodesIds,
+                    nodes,
+                    weightedRoutingMetadata.getWeightedRouting(),
+                    getWeightedRoutingDefaultWeight()
+                );
+            case LOCAL:
+                return indexShard.preferNodeActiveInitializingShardsWeightedIt(
+                    Collections.singleton(localNodeId),
+                    nodes,
+                    weightedRoutingMetadata.getWeightedRouting(),
+                    getWeightedRoutingDefaultWeight()
+                );
+            case ONLY_LOCAL:
+                if (!indexShard.isNodeWeighedAway(
+                    weightedRoutingMetadata.getWeightedRouting(),
+                    nodes,
+                    localNodeId,
+                    getWeightedRoutingDefaultWeight()
+                )) return indexShard.onlyNodeActiveInitializingShardsIt(localNodeId);
+                else {
+                    throw new PreferenceBasedSearchNotAllowedException(
+                        "Preference based routing not allowed on weigh away node with strict weighted shard routing enabled"
+                    );
+                }
+            case ONLY_NODES:
+                String nodeAttributes = preference.substring(Preference.ONLY_NODES.type().length() + 1);
+                return indexShard.onlyNodeSelectorActiveInitializingWeightedShardsIt(
+                    nodeAttributes.split(","),
+                    nodes,
+                    weightedRoutingMetadata.getWeightedRouting(),
+                    getWeightedRoutingDefaultWeight()
```

Comment:
There could be multiple cases for each of these cases. Lets add comments on the behavioural change. We might also want to document the same